janitor/functions/case_when.py

```diff
@@ -99,100 +102,97 @@ def case_when(df: pd.DataFrame, *args, column_name: str) -> pd.DataFrame:
         `result` can be a scalar, a 1-D array, or a callable.
         If `result` is a callable, it should evaluate to a 1-D array.
         For a 1-D array, it should have the same length as the DataFrame.
-        The `default` argument applies if none of `condition0`,
-        `condition1`, ..., evaluates to `True`.
-        Value can be a scalar, a callable, or a 1-D array. if `default` is a
-        callable, it should evaluate to a 1-D array.
+    :param default: scalar, 1-D array or callable.
+        If callable, it should evaluate to a 1-D array.
         The 1-D array should be the same length as the DataFrame.
+        The element inserted in the output when all conditions
+        evaluate to False.
```

Comment:
My only concern is that this is a breaking change which probably warrants a deprecation cycle: users who depended on providing `2n+1` args (with the last arg treated as the "default"), will now be met with a TypeError due to a missing `default` kwarg.

I'm thinking if a better solution here is to (1) make `default` default to some value like `"__pyjanitor_default__"` (it can't be None since I suppose None is a valid default value), (2) introduce a DeprecationWarning to when we detect `default` is the special value from step1. Point the user towards using the explicit `default` kwarg. while manually popping the last arg into `default`. (3) Perform the remaining checks as if the user had passed in `default`, i.e., if `len(args)` is odd, then raise error.

Thoughts?

Reply:
Been a while @thatlittleboy hope you are well and good. 

I'm thinking of closing this PR and leave things as they are; the core is not really changed, maybe a future PR can improve performance. But yea, the breaking change is not worth it

Reply:
Hey @samukweku , thanks, it's been a while. Hoping to jump back in some time soon to assist more, probably during the holiday season when work's a little lighter.

Back on topic: I do like the idea of specifying the `default` kwarg though, I think it's a step in the right direction. I've always found the (2n+1) "API" for `case_when` a little confusing to read when written out as code. The breaking change just needs a little careful handling.

I'll suggest keeping the PR open instead of closing for now, someone else can pick up from where you left off (I'll volunteer, if I can find the time :x ).

Reply:
@thatlittleboy made some edits to the code to cater for deprecation. Would love your feedback on this. @ericmjl @Zeroto521 pls feel free to chime in when you have some bandwidth